src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+                               int assigneeID) {
+    // Create the StatementAccessPair of the assignment with the new and old variables
+    std::shared_ptr<FieldAccessExpr> fa_assignee =
+        std::make_shared<FieldAccessExpr>(instantiation_->getNameFromAccessID(assigneeID));
+    std::shared_ptr<FieldAccessExpr> fa_assignment =
+        std::make_shared<FieldAccessExpr>(instantiation_->getNameFromAccessID(assignmentID));
+    std::shared_ptr<AssignmentExpr> assignmentExpression =
+        std::make_shared<AssignmentExpr>(fa_assignment, fa_assignee, "=");
+    std::shared_ptr<ExprStmt> expAssignment = std::make_shared<ExprStmt>(assignmentExpression);
+    std::shared_ptr<Statement> assignmentStatement =
+        std::make_shared<Statement>(expAssignment, nullptr);
+    std::shared_ptr<StatementAccessesPair> pair =
+        std::make_shared<StatementAccessesPair>(assignmentStatement);
+    std::shared_ptr<Accesses> newAccess = std::make_shared<Accesses>();
+    newAccess->addWriteExtent(assignmentID, Extents({0, 0, 0}));
+    newAccess->addReadExtent(assigneeID, Extents({0, 0, 0}));
+    pair->setAccesses(newAccess);
+    domethod->getStatementAccessesPairs().push_back(pair);
+
+    // Add the new expressions to the map
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignment, assignmentID);
+    instantiation_->getExprToAccessIDMap().emplace(fa_assignee, assigneeID);
+  }
+
+  /// @brief checks if there is a read operation before the first write operation in the given
+  /// multistage.
+  /// @param[in]    id  original FieldAccessID of the field to check
+  /// @return           true if there is a read-access before the first write access
+  bool checkReadBeforeWrite(int id) {
+    for(auto stageItGlob = multiStagePrt_->getStages().begin();
+        stageItGlob != multiStagePrt_->getStages().end(); ++stageItGlob) {
+      DoMethod& doMethod = (**stageItGlob).getSingleDoMethod();
+      for(int stmtIndex = 0; stmtIndex < doMethod.getStatementAccessesPairs().size(); ++stmtIndex) {
+        const std::shared_ptr<StatementAccessesPair>& stmtAccessesPair =
+            doMethod.getStatementAccessesPairs()[stmtIndex];
+
+        // Find first if this statement has a read
+        auto readAccessIterator = stmtAccessesPair->getCallerAccesses()->getReadAccesses().find(id);
+        if(readAccessIterator != stmtAccessesPair->getCallerAccesses()->getReadAccesses().end()) {
+          return true;
+        }
+        // If we did not find a read statement so far, we have  a write first and do not need to
+        // fill the cache
+        auto wirteAccessIterator =
+            stmtAccessesPair->getCallerAccesses()->getWriteAccesses().find(id);
+        if(wirteAccessIterator != stmtAccessesPair->getCallerAccesses()->getWriteAccesses().end()) {
+          return false;
+        }
+      }
+    }
+    return false;
+  }
+
+  bool checkReadOnlyAccess(int id) {
```

Comment:
It would be better to stick to the convention of calling it `AccessID` instead of just `id`.

The formatting is really broken in this function